devito/finite_differences/differentiable.py

```diff
@@ -186,20 +187,93 @@ def _has(self, pattern):
         return super(Differentiable, self)._has(pattern)
 
 
-class Add(sympy.Add, Differentiable):
-    pass
+class DifferentiableOp(Differentiable):
 
+    def __new__(cls, *args, **kwargs):
+        obj = cls.__base__.__new__(cls, *args, **kwargs)
 
-class Mul(sympy.Mul, Differentiable):
-    pass
+        # Unfortunately SymPy may build new sympy.core objects (e.g., sympy.Add),
+        # so here we have to rebuild them as devito.core objects
+        obj = diffify(obj)
 
+        return obj
 
-class Pow(sympy.Pow, Differentiable):
-    pass
 
+class Add(sympy.Add, DifferentiableOp):
+    __new__ = DifferentiableOp.__new__
 
-class Mod(sympy.Mod, Differentiable):
-    pass
+
+class Mul(sympy.Mul, DifferentiableOp):
+    __new__ = DifferentiableOp.__new__
+
+
+class Pow(sympy.Pow, DifferentiableOp):
+    __new__ = DifferentiableOp.__new__
+
+
+class Mod(sympy.Mod, DifferentiableOp):
+    __new__ = DifferentiableOp.__new__
+
+
+class diffify(object):
+
+    """
+    Helper class based on single dispatch to reconstruct all nodes in a sympy
+    tree such they are all of type Differentiable.
+
+    Notes
+    -----
+    The name "diffify" stems from SymPy's "simpify", which has an analogous task --
+    converting all arguments into SymPy core objects.
+    """
+
+    def __new__(cls, obj):
+        args = [diffify._doit(i) for i in obj.args]
```

Comment:
would move that inside `_doit`

Reply:
yes, I think I can do this

Reply:
no, turns out I cannot. I need the first lines of `_doit` to be executed for each `i` (ie for each obj argument), and I don't want to duplicate code